cli/sql: handle protocol errors caused by COPY in a user-friendly manner

knz opened the pull request:
Prior to this patch, issuing COPY on the SQL CLI would cause
a non-recoverable protocol error with an unscrutable error message:

```
root@:26257/test> copy kv from stdin;
pq: unknown response for simple query: 'G'
error retrieving the transaction status: driver: bad connection
connection lost; opening new connection: all session settings will be lost
root@:26257/ ?>
```

This patch enhances UX by providing a friendly error message instead:

```
root@:26257/test> copy kv from stdin;
woops! COPY has confused this client! Suggestion: use 'psql' for COPY.
connection lost; opening new connection: all session settings will be lost
root@:26257/test>
```

Fixes #17577.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17607)
<!-- Reviewable:end -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thank you!

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17607#-:-KrWE4L2ryM2f3z7lvt5:bro3093)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/cli/sql_util.go, line 459 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17607#-KrWsO2xpkLzBGAzRIZz:-KrWsO2xpkLzBGAzRI_-:b-ugbd0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/881d69b7d9dd0efd82395817dd28dc8ec62c4c7d/pkg/cli/sql_util.go#L459)):*
> ```Go
> 	// it.
> 	conn.Close()
> 	if err := conn.ensureConn(); err != nil {
> ```

This is unfortunate because it will clear the session settings of the user without their knowledge. I'm ok with keeping the code as is because we think adding that text to the error message is rare, but we could consider adding `; connection reset` to the error message to further inform users what happened.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17607)*
<!-- Sent from Reviewable.io -->

mjibson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17607#-:-KrWsnXlwC7Ou7Sdhvtp:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz committed:
cli/sql: handle protocol errors caused by COPY in a user-friendly manner

Prior to this patch, issuing COPY on the SQL CLI would cause
a non-recoverable protocol error with an unscrutable error message:

```
root@:26257/test> copy kv from stdin;
pq: unknown response for simple query: 'G'
error retrieving the transaction status: driver: bad connection
connection lost; opening new connection: all session settings will be lost
root@:26257/ ?>
```

This patch enhances UX by providing a friendly error message instead:

```
root@:26257/test> copy kv from stdin;
woops! COPY has confused this client! Suggestion: use 'psql' for COPY.
connection lost; opening new connection: all session settings will be lost
root@:26257/test>
```

knz commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/cli/sql_util.go, line 459 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17607#-KrWsO2xpkLzBGAzRIZz:-KsERj_1nWQklSNSvTxS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/881d69b7d9dd0efd82395817dd28dc8ec62c4c7d/pkg/cli/sql_util.go#L459)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

This is unfortunate because it will clear the session settings of the user without their knowledge. I'm ok with keeping the code as is because we think adding that text to the error message is rare, but we could consider adding `; connection reset` to the error message to further inform users what happened.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17607)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYR!
